EMULATE_FUNCTION_POINTER_CASTS with WebAssembly should not disable async compilation

I was recently trying to asynchronously load my compiled WebAssembly. I used ``Module.instantiateWasm``, but kept running into issues such as "real__malloc is not callable".

After much hunting around in the emscripten source code I found that, for asynchronous loading of WebAssembly to work, you need to compile with "-s BINARYEN_ASYNC_COMPILATION=1" or "-s SWAPPABLE_ASM_MODULE=1".

I suggest updating the documentation for ``Module.instantiateWasm`` to mention these compilation flags, to avoid others having to go through the same pain.

Documentation URL: https://kripken.github.io/emscripten-site/docs/api_reference/module.html#Module.instantiateWasm

in my own project we need this too, and after making the same patch as tbfleming, i found that some of the binaryen tests were failing. Most of them fails because of the same reason, and its demonstated by test_legalize_js_ffi , with error -> AssertionError: i64 not converted to i32 in exports.
seems like its similar to this issue https://github.com/kripken/emscripten/issues/5370. any idea where it might be causing the problem?

--- a/emcc.py
+++ b/emcc.py
@@ -1304,8 +1304,8 @@ There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR P
         if 'MEM_INIT_METHOD' in settings_changes:
           exit_with_error('MEM_INIT_METHOD is not supported in wasm. Memory will be embedded in the wasm binary if threads are not used, and included in a separate file if threads are used.')
         options.memory_init_file = True
-        # async compilation requires wasm-only mode, and also not interpreting (the interpreter needs sync input)
-        if shared.Settings.BINARYEN_ASYNC_COMPILATION == 1 and shared.Building.is_wasm_only() and 'interpret' not in shared.Settings.BINARYEN_METHOD:
+        # async compilation requires not interpreting (the interpreter modes needs sync input)
+        if shared.Settings.BINARYEN_ASYNC_COMPILATION == 1 and 'interpret' not in shared.Settings.BINARYEN_METHOD:
           # async compilation requires a swappable module - we swap it in when it's ready
           shared.Settings.SWAPPABLE_ASM_MODULE = 1
         else:

--- a/tests/test_browser.py
+++ b/tests/test_browser.py
@@ -3704,7 +3704,7 @@ window.close = function() {
       (['-O3'], 1),
       (['-s', 'BINARYEN_ASYNC_COMPILATION=1'], 1), # force it on
       (['-O1', '-s', 'BINARYEN_ASYNC_COMPILATION=0'], 0), # force it off
-      (['-s', 'BINARYEN_ASYNC_COMPILATION=1', '-s', 'BINARYEN_METHOD="native-wasm,asmjs"'], 0), # try to force it on, but have it disabled
+      (['-s', 'BINARYEN_ASYNC_COMPILATION=1', '-s', 'BINARYEN_METHOD="interpret-binary"'], 0), # try to force it on, but have it disabled
     ]:
       print(opts, expect)
       self.btest('binaryen_async.c', expected=str(expect), args=common_args + opts)

--- a/tests/test_other.py
+++ b/tests/test_other.py
@@ -2915,6 +2915,7 @@ myreade(){
            '-o', 'proxyfs_test.js', 'proxyfs_test.c',
            '--embed-file', 'proxyfs_embed.txt', '--pre-js', 'proxyfs_pre.js',
            '-s', 'EXTRA_EXPORTED_RUNTIME_METHODS=["ccall", "cwrap"]',
+           '-s', 'BINARYEN_ASYNC_COMPILATION=0',
            '-s', 'MAIN_MODULE=1']).communicate()
     # Following shutil.copyfile just prevent 'require' of node.js from caching js-object.
     # See https://nodejs.org/api/modules.html
@@ -7851,8 +7852,8 @@ int main() {
 
   def test_binaryen_warn_sync(self):
     if SPIDERMONKEY_ENGINE not in JS_ENGINES: self.skipTest('cannot run without spidermonkey')
-    # using a fallback to asm.js will disable async
-    for method in ['native-wasm,asmjs', 'native-wasm', None]:
+    # interpreting will disable async
+    for method in ['interpret-binary', 'native-wasm', None]:
       cmd = [PYTHON, EMCC, path_from_root('tests', 'hello_world.cpp'), '-s', 'WASM=1']
       if method is not None:
         cmd += ['-s', 'BINARYEN_METHOD="' + method + '"']
@@ -7860,7 +7861,7 @@ int main() {
       err = run_process(cmd, stdout=PIPE, stderr=PIPE).stderr
       print(err)
       warning = 'BINARYEN_ASYNC_COMPILATION disabled due to user options. This will reduce performance and compatibility'
-      if method and ',' in method:
+      if method and 'interpret' in method:
         self.assertContained(warning, err)
       else:
         self.assertNotContained(warning, err)

